evadb/third_party/vector_stores/milvus.py

```diff
@@ -0,0 +1,149 @@
+            user=milvus_user,
+            password=milvus_password,
+            db_name=milvus_db_name,
+            token=milvus_token,
+        )
+
+    return _milvus_client_instance
+
+
+class MilvusVectorStore(VectorStore):
+    def __init__(self, index_name: str) -> None:
+        # Milvus URI is the only required
+        self._milvus_uri = ConfigurationManager().get_value("third_party", "MILVUS_URI")
+
+        if not self._milvus_uri:
+            self._milvus_uri = os.environ.get("MILVUS_URI")
+
+        assert (
+            self._milvus_uri
+        ), "Please set your Milvus URI in evadb.yml file (third_party, MILVUS_URI) or environment variable (MILVUS_URI)."
+
+        # Check other Milvus variables for additional customization
+        self._milvus_user = ConfigurationManager().get_value(
+            "third_party", "MILVUS_USER"
+        )
+
+        if not self._milvus_user:
+            self._milvus_user = os.environ.get("MILVUS_USER", "")
+
+        self._milvus_password = ConfigurationManager().get_value(
+            "third_party", "MILVUS_PASSWORD"
+        )
+
+        if not self._milvus_password:
+            self._milvus_password = os.environ.get("MILVUS_PASSWORD", "")
+
+        self._milvus_db_name = ConfigurationManager().get_value(
+            "third_party", "MILVUS_DB_NAME"
+        )
+
+        if not self._milvus_db_name:
+            self._milvus_db_name = os.environ.get("MILVUS_DB_NAME", "")
+
+        self._milvus_token = ConfigurationManager().get_value(
+            "third_party", "MILVUS_TOKEN"
+        )
+
+        if not self._milvus_token:
+            self._milvus_token = os.environ.get("MILVUS_TOKEN", "")
+
+        self._client = get_milvus_client(
+            milvus_uri=self._milvus_uri,
+            milvus_user=self._milvus_user,
+            milvus_password=self._milvus_password,
+            milvus_db_name=self._milvus_db_name,
+            milvus_token=self._milvus_token,
+        )
+        self._collection_name = index_name
+
+    def create(self, vector_dim: int):
+        if self._collection_name in self._client.list_collections():
+            self._client.drop_collection(self._collection_name)
+        self._client.create_collection(
+            collection_name=self._collection_name,
+            dimension=vector_dim,
+            metric_type="COSINE",
+        )
+
+    def add(self, payload: List[FeaturePayload]):
+        milvus_data = [
+            {
+                "id": feature_payload.id,
+                "vector": feature_payload.embedding.reshape(-1).tolist(),
+            }
+            for feature_payload in payload
+        ]
+        ids = [feature_payload.id for feature_payload in payload]
+
+        # Milvus Client does not have upsert operation, perform delete + insert to emulate it
+        self._client.delete(collection_name=self._collection_name, pks=ids)
+
+        self._client.insert(collection_name=self._collection_name, data=milvus_data)
```

Comment:
What happens if we directly call insert on existing ids?

Reply:
Can this not be used ? https://milvus.io/docs/upsert_entities.md

Reply:
There is no `upsert` for the Milvus client object itself. Just inserting on existing ids will not update it.

Reply:
Had the same question. Can we get collection from the client and upsert using collection APIs? Maybe I missed something here.

Reply:
There seems to be no methods in the client to get the collection objects, just the collection names through `list_collections`. Maybe something like this could work:
```python
client_connection_alias = self._client._using
collection = Collection(self._collection_name, using=client_connection_alias)
collection.upsert(milvus_data)
```
My biggest concern with this though is the use of the private `_using` variable in the client.